docs/man/greydbd.8
.TH GREYDBD 8 "1.0" "May 30, 2012" ""
.SH NAME
.B greydbd
\- spamilter greylist support database service
.SH SYNOPSIS
greydbd [-v] [-d] [-p port number] [-h psql hostname] [-i psql host port] [-j psql database device name] [-k psql database user name] [-l psql database user password]
.br
.SH OPTIONS
.TP
.B -v
display version
.TP
.B -p
service port number
.TP
.B -h
psql hostname
.TP
.B -i
psql host port
.TP
.B -j
psql database device name
.TP
.B -k
psql database user name
.TP
.B -l
psql database user password
.SH DESCRIPTION
provides a database storage and aging mechanisim to track ipaddress and email tuples.
.br
.SH VERSION
This documentation describes
.B greydbd
version 1.0
.SH "SEE ALSO"
spamilter(8), ipfwmtad(8)
.br
.I http://www.wanlink.com/spamilter/ site
.SH AUTHOR
.br
Neal Horman <spamilter@wanlink.com>


Changelog
Changelog;
	20150129
	* mod - be more aggresive about ipfw blocking MTAs when in blacklist type of conditions
	* fix - don't double ipfw inculpate ip address

	20150125
	* mod - configure - do some clean up / modernization
	* fix - segfault in badext.c when logging 'unable to open file'
	* mod - change 'configure --sendmail-dir' can work on FreeBSD
	* mod - change 'configure --with-libspf' back to previous behavior, and split builtin behavior out to '--with-libspf-builtin'
	* add - 'make install-config' target to do initial config file installation
	* add - support for regex pattern matching for BWLIST mbox

	20150124
	* mod - mlfi_regex_mboxsplit() add '=' to valid characters list of mbox regex pattern
	* mod - fix dnsbl regression

	20150123
	* add - dns, dbl and dnsbl changes to grok ipv6

	20150121
	* mod - add ipv6 lookups for MtaHostChk functionality

	* tag - build_140227
	* mod - un-nest callback functions for llvm compatiblity

	* tag - build_130713_1
	* add - HLO host DBL checking
	* add - body host DBL checking

	* tag - build_121118_1
	* add - session id for logging purposes

	* tag - build_121006
	* add - do geoip checking for 'X-Originating-IP' hotmail header

	* fix - don't ignore SMFIS result from BodyHost checking function
	* mod - tweak replyto BWL handling behavior

	* mod - don't do Body URL black listing and Body URL CC black listing if there is rcpt or sndr Accept
		action, or is localnet host
	* mod - reply-to header delivery checking to ignore '<>' return path
	* add - setproctitle for "watcher" and "worker"

	* tag - build_120625_2
	* fix - regression of dns_query_rr_a() function use by "Invalid MTA hostname" filter

	* tag - build_120625_1
	* mod - bwlist to support postgresql
	* mod - table driver framework to support postgresql table driver
	* add - add postgresql table driver

	* tag - build_120618_1
	* mod - rewrite bwlist to be based off of the table framework and the flat file table driver
	* add - add flat file table driver
	* add - generic table driver framework

	* add - filter - using regex, scan the body for http/https/ftp urls and do BWL and geoip contry code lookup rejection

	* add - filter - received by/from header hostname BWL checking and geoip contry code ip lookup rejection
	* mod - change all injected headers to use the basic format of "X-Milter: Spamilter DataSet=xxx; foo=bar; foo=bar; ..."
	* add - filter - BWL and geoip country code lookup rejection for connecting mta

	* mod - start working towards ipv6 support

	* add - filter - reply to header return address verification
	* mod - test all RBL entries for the connecting mta ip address, to inject headers when tagging instead of rejecting

	* add - filter - grey listing mechanism (requires postgresl)
	* add - filter - fwdhost lookup - allow an edge MTA to consult an athorative inner MTA for recipent delivery status

	* add - ability for local definitions of localnet testing in ifi_islocalnet() via addtions to the struct _localnethosts
		defintion from the ifilocal.inc file
	* add - framework for local content filtering via c code from hndlrs_contentfilter_local.inc file
	* add - framework for local message duplication based on sender or recipient via hard coded rules in c code from
		dupelocal.inc file
	* mod - treat SASL authorized connections as a localnet host

	* add - filter to reject if recipent is not a target listed in virtusertable.db
	* add - filter to reject if recipent is not a target listed in aliases.db
	* add - filter to reject if recipent is not a local system user
	* mod - don't do dns lookups on a given rbl host until the reject/tag stage as listed/configured in spamilter.rc
		instead of doing the lookups at connect and then reject/tagging at stage
	* fix - local asprintf implementation in nstring.c to not segfault on a %s parameter when NULL
	* fix - null pointer sanity tests
	* add - cli -f to spamilter to keep it in foreground with debug info
	* mod - to EOM handler to tag or flag with all filter results instead of just one result
	* fix - for sender address verification to treat the mbox in a case insensative way per RFC2821
	* mod - don't collect body if not used
	* fix - debugmode > 1 works now

version - 0.60 - 051127
	* fix - actually use the configured db path instead of the hard coded one - patch from Luns Tee
	* fix - make sure the priv->smtprc is initialized. - patch from Luns Tee
	* fix - remove NULL assignment of argptr, Linux 7.? barfs on it
	* fix - Makefile.tmpl so that SunOs CC can deal with += vars, add -lpthread for SunOs
	* fix - memory leak in ifi.c - patch from Luns Tee
	* fix - memory leak in dnsbl.c - patch from Luns Tee
	* mod - to combat rfc ignorant mta hosts, try twice to negotiate an accepted
		"mail from:, rcpt to:" sequence, first with <>, then with <postmaster@`local host name`>
		if that fails.
	* fix - globally use mx_res_search that resizes the query buffer as needed - drastically altered
		patch from Luns Tee
	* mod - if not doing attachment checks, don't do body processing - patch from Luns Tee
	* mod - add --with-???? options to configure. see configure --help for more details
	* mod - use res_n???? thread safe resolver library calls
	* mod - rename spf.[ch] to spfapi.[ch] so as to not conflict with libspf's spf.h
	* mod - drop the included libspf in favor of using the "current" version as published on the
		web site at http://www.libspf.org now that --with-libspf is available in configure
		ntbw. installation of the library is not required, rather just configuration and
		compilation, as it is statically linked.
	* mod - ipfwmtad to handle exculpate correctly
	* add - ipfwmtad -u ipfw rule number
	* add - initial support for OpenBSD
	* mod - move MsExtChk core code into separate source module
	* add - support for reading MsExtChk file list from db.extensions in lue of being hard coded
	* add - support for ipfw2 in ipfwmtad_direct
	* mod - compile with Wall by default

src/greydbd.c
/*--------------------------------------------------------------------*
 *
 * Developed by;
 *	Neal Horman - http://www.wanlink.com
 *	Copyright (c) 2009 Neal Horman. All Rights Reserved
 *
 *	Redistribution and use in source and binary forms, with or without
 *	modification, are permitted provided that the following conditions
 *	are met:
 *	1. Redistributions of source code must retain the above copyright
 *	   notice, this list of conditions and the following disclaimer.
 *	2. Redistributions in binary form must reproduce the above copyright
 *	   notice, this list of conditions and the following disclaimer in the
 *	   documentation and/or other materials provided with the distribution.
 *	3. All advertising materials mentioning features or use of this software
 *	   must display the following acknowledgement:
 *	This product includes software developed by Neal Horman.
 *	4. Neither the name Neal Horman nor the names of any contributors
 *	   may be used to endorse or promote products derived from this software
 *	   without specific prior written permission.
 *	
 *	THIS SOFTWARE IS PROVIDED BY NEAL HORMAN AND ANY CONTRIBUTORS ``AS IS'' AND
 *	ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 *	IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 *	ARE DISCLAIMED.  IN NO EVENT SHALL NEAL HORMAN OR ANY CONTRIBUTORS BE LIABLE
 *	FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 *	DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 *	OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 *	HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 *	LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 *	OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 *	SUCH DAMAGE.
 *
 *	CVSID:  $Id: greydbd.c,v 1.11 2013/01/04 02:37:49 neal Exp $
 *
 * DESCRIPTION:
 *	application:	greydbd
 *	module:		greydbd.c
 *--------------------------------------------------------------------*/

static char const cvsid[] = "@(#)$Id: greydbd.c,v 1.11 2013/01/04 02:37:49 neal Exp $";

#include "config.h"

#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <signal.h>
#include <termios.h> 
#include <fcntl.h>
#include <string.h>
#include <syslog.h>
#include <stdarg.h>
#include <time.h>
#include <ctype.h>

#include <sys/time.h>
#include <sys/ioctl.h>
#include <sys/wait.h>
#include <sys/stat.h>
#include <sys/types.h>
#include <netinet/in.h>
#include <netinet/tcp.h>
#include <sys/socket.h>
#include <arpa/inet.h>

#include "md5api.h"


#include "inet.h"
#include "pgsql.h"

int		gDebug		= 0;
int		gSd		= INVALID_SOCKET;
ppgsql_srvr	gpDbsrvr	= NULL;

char const	*gpDbHost	= "localhost";
char const	*gpDbHostPort	= "5432";
char const	*gpDbDevice	= "spamilter";
char const	*gpDbUserName	= "spamilter";
char const	*gpDbUserPass	= "";

ppgsql_srvr dbinit()
{

	gpDbsrvr = pgsql_srvr_destroy(gpDbsrvr,1);
	gpDbsrvr = pgsql_srvr_init(gpDbHost,gpDbHostPort,gpDbDevice,gpDbUserName,gpDbUserPass);
	if(gpDbsrvr != NULL)
		pgsql_open(gpDbsrvr);

	return gpDbsrvr;
}

int hashIsWhite(char *phash, unsigned int hashlen)
{	unsigned long count = 0;

	if(phash != NULL && hashlen != 0)
	{
		if(gpDbsrvr == NULL || !pgsql_IsOpen(gpDbsrvr->dbl))
			dbinit();

		if(gpDbsrvr != NULL && gpDbsrvr->dbl != NULL)
		{	ppgsql_query dblq = pgsql_exec_printf(gpDbsrvr->dbl,"select greylist('%*.*s')",hashlen,hashlen,phash);
			int dbreset = 1;

			if(dblq != NULL)
			{
				if(dblq->ok)
				{
					if(dblq->numRows > 0)
					{	char buf[128];

						memset(buf,0,sizeof(buf));
						if(pgsql_getFieldBuf(dblq,0,0,buf,sizeof(buf)-1))
							count = atol(buf);
					}
					dbreset = 0;
				}
				else
					syslog(LOG_DEBUG,"hashIsWhite: dblq %c= NULL && dbl->ok %u && dblq->numRows %u\n"
						,(dblq != NULL ? '=' : '!')
						,(dblq != NULL ? dblq->ok : 0)
						,(dblq != NULL ? dblq->numRows : 0)
						);

				pgsql_freeResult(dblq);
			}

			if(dbreset)
			{
				syslog(LOG_INFO,"hashIsWHite: psql_srvr_destroy");
				gpDbsrvr = pgsql_srvr_destroy(gpDbsrvr,1);
			}
		}
	}

	return (count > 1);
}

void processSocket(int sd)
{	char sockbuf[1024];

	memset(&sockbuf,0,sizeof(sockbuf));
	if(NetSockSelectOne(sd,10))
	{	unsigned long peerIp = 0;
		unsigned short peerPort = 0;
		int rc = NetSockRecvFrom(sd,sockbuf,sizeof(sockbuf), &peerIp, &peerPort, NULL);

		if(rc != SOCKET_ERROR && peerIp == 0x7f000001)
		{
			int i;
			char hashbuf[MD5_DIGEST_STRING_LENGTH];
			char *pStrHashResult[] = {"ok", "fail"};
			int hashIndex = 0;

			for(i=0; i<rc; i++)
				sockbuf[i]=tolower(sockbuf[i]);

			memset(&hashbuf,0,sizeof(hashbuf));
			hashIndex = !hashIsWhite(MD5Hash(sockbuf,rc,hashbuf),MD5_DIGEST_STRING_LENGTH-1);
			NetSockPrintfTo(sd,peerIp,peerPort,"<%s>",pStrHashResult[hashIndex]);

			syslog(LOG_INFO,"%*.*s = %*.*s - %s - %u.%u.%u.%u:%u '<%s>'\n"
				,rc,rc,sockbuf
				,MD5_DIGEST_STRING_LENGTH-1,MD5_DIGEST_STRING_LENGTH-1,hashbuf
				,(hashIndex ? "grey" : "white")
				,(unsigned int)(peerIp>>24)&0xff,(unsigned int)(peerIp>>16)&0xff,(unsigned int)(peerIp>>8)&0xff,(unsigned int)peerIp&0xff,peerPort
				,pStrHashResult[hashIndex]
				);
		}
	}
}

void childShutdown();

void child_signal_hndlr(int signo)
{
	switch(signo)
	{
		case SIGHUP:
			syslog(LOG_ERR,"Signal %u received\n",signo);
			break;
		case SIGTERM:
		case SIGQUIT:
		case SIGINT:
		case SIGSEGV:
			syslog(LOG_ERR,"Shutdown with signal %u\n",signo);
			childShutdown();
			exit(signo);
			break;
	}
}

void childShutdown()
{
	NetSockClose(&gSd);
	if(gpDbsrvr != NULL)
		gpDbsrvr = pgsql_srvr_destroy(gpDbsrvr,1);
	closelog();
}

int childMain(short port)
{	int	quit	= 0;
	int	rc = -1;

	signal(SIGTERM,child_signal_hndlr);
	signal(SIGQUIT,child_signal_hndlr);
	signal(SIGINT,child_signal_hndlr);
	signal(SIGHUP,child_signal_hndlr);
	signal(SIGSEGV,child_signal_hndlr);

	syslog(LOG_ERR,"Child: Started using listening port '%u'",port );

	gSd = NetSockOpenUdp(0x7f000001,port);
	dbinit();
	if(gSd != INVALID_SOCKET && gpDbsrvr != NULL)
	{
		while(!quit)
		{
			processSocket(gSd);
			usleep(10000);
		}

		rc = 0;
		syslog(LOG_ERR,"Child: Exiting: normal");
	}

	if(gSd == INVALID_SOCKET)
		syslog(LOG_ERR,"Child: Exiting: Invalid Tcp socket");
	if(gpDbsrvr == NULL)
		syslog(LOG_ERR,"Childe: exiting: NULL pgsql connection");

	childShutdown();

	return(rc);
}

void childStart(short port, int count, int forked)
{	int	flags = 0;

	if(forked)
	{
		close(STDIN_FILENO);
		close(STDOUT_FILENO);
		close(STDERR_FILENO);
	}
	else
		flags = LOG_PERROR;

	while(count--)
	{
		openlog("greydbd",flags|LOG_NDELAY|LOG_PID,LOG_DAEMON);
		if(childMain(port) == -1)
		{
			syslog(LOG_ERR,"Error starting child, sleeping 30 seconds, %u retries left.",count);
			sleep(30000);
		}
	}
}

void showVersion()
{
	printf("greydbd - Version 1.0 - spamilter greylist database support service\n");
}

void showUsage()
{
	showVersion();
	printf("usage:\n"
		"\t-v = Service Version Information\n"
		"\t-d = Service Debug Mode\n"
		"\t-p = Serivce Port Number\n"
		"\t-h = PostgreSql Host Name\n"
		"\t-i = PostgreSql Host Port\n"
		"\t-j = PostgreSql DataBase Device Name\n"
		"\t-k = PostgreSql DataBase User Name\n"
		"\t-l = PostgreSql DataBase User Password\n"
		);
}

int main(int argc, char **argv)
{	int	count = 3;
	short	port = (short)7892;
	int	forking = 1;
	char	opt;
	char	*optflags = "d:p:h:i:j:k:l:v?";

	while((opt = getopt(argc,argv,optflags)) != -1)
	{
		switch(opt)
		{
			case 'd':
				if(optarg != NULL && *optarg)
					forking = 1 - ((gDebug = atoi(optarg))>0);
				if(gDebug)
					printf("debug mode\n");
				break;

			/* server mode stuff */
			case 'p':
				if(optarg != NULL && *optarg)
					port = (short)atoi(optarg);
				break;
			case 'h':
				if(optarg != NULL && *optarg)
					gpDbHost = strdup(optarg);
				break;
			case 'i':
				if(optarg != NULL && *optarg)
					gpDbHostPort = strdup(optarg);
				break;
			case 'j':
				if(optarg != NULL && *optarg)
					gpDbDevice = strdup(optarg);
				break;
			case 'k':
				if(optarg != NULL && *optarg)
					gpDbUserName = strdup(optarg);
				break;
			case 'l':
				if(optarg != NULL && *optarg)
					gpDbUserPass = strdup(optarg);
				break;
			case 'v':
				showVersion();
				return 0;
				break;
			case '?':
			default:
				showUsage();
				return 1;
				break;
		}
	}
	argc -= optind;
	argv += optind;

	if(forking)
	{
		if(fork() == 0)
			childStart(port,count,1);
	}
	else
		childStart(port,count,0);

	return 0;
}
